daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/ExplicitDisclosureTest.scala

```diff
@@ -0,0 +1,736 @@
+                    """,
+                    Array(
+                      SContractId(contractId),
+                      SValue.SText(testKeyName),
+                      SValue.SParty(disclosureParty),
+                    ),
+                    committers = Set(disclosureParty),
+                    disclosedContracts = ImmArray(disclosedContract),
+                  )
+
+                inside(result) { case Left(SError.SErrorDamlException(ContractKeyNotFound(key))) =>
+                  key shouldBe buildContractKey(disclosureParty)
+                  events shouldBe Array("contractById queried")
+                }
+              }
+            }
+          }
+
+          "looking up contract keys" - {
+            "fail to evaluate lookup contract keys" in {
+              val (result, events) =
+                evaluateExprApp(
+                  e"""\(contractId: ContractId TestMod:House) (label: Text) (maintainer: Party) ->
+                          ubind ignore: Unit <- TestMod:exercise contractId
+                          in TestMod:lookup_by_key label (Some @Party maintainer)
+                  """,
+                  Array(
+                    SContractId(contractId),
+                    SValue.SText(testKeyName),
+                    SValue.SParty(ledgerParty),
+                  ),
+                  committers = Set(ledgerParty),
+                  getKey = Map(
+                    GlobalKeyWithMaintainers(ledgerContractKey, Set(ledgerParty)) -> contractId
+                  ),
+                  getContract = Map(contractId -> ledgerContract),
+                )
+
+              inside(result) { case Right(SValue.SOptional(None)) =>
+                events shouldBe Array("contractById queried", "getContract queried")
+              }
+            }
+
+            "fail to lookup disclosed contract keys" in {
+              forAll(disclosedContracts) { case (disclosedContract, _) =>
+                val (result, events) =
+                  evaluateExprApp(
+                    e"""\(contractId: ContractId TestMod:House) (label: Text) (maintainer: Party) ->
+                          ubind ignore: Unit <- TestMod:exercise contractId
+                          in TestMod:lookup_by_key label (Some @Party maintainer)
+                    """,
+                    Array(
+                      SContractId(contractId),
+                      SValue.SText(testKeyName),
+                      SValue.SParty(disclosureParty),
+                    ),
+                    committers = Set(disclosureParty),
+                    disclosedContracts = ImmArray(disclosedContract),
+                  )
+
+                inside(result) { case Right(SValue.SOptional(None)) =>
+                  events shouldBe Array("contractById queried")
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+
+    "with contracts stored in multiple locations" - {
```

Comment:
From what I can tell, we run the same set of tests in 3 different versions:
1. With only `getContract/getKey` set.
2. With only `disclosedContracts` set.
3. With both set.

Can we factor this out via TableDrivenProperties or something? This just seems to be asking for things to get subtly out of sync.

Reply:
Addressed this issue by refactoring common test code into a set of test assertion methods.